### ISSUE TITLE
Inconsistent behavior between memory and redis

### ISSUE THREAD

**Issue by akornatskyy** [slact/nchan#391]
nginx: 1.12.1
nchan: 1.1.5 - 1.1.7
redis: 4.0.1

I use [http://www.websocket.org/echo.html](http://www.websocket.org/echo.html) to send a message to /pubsub endpoint. Here is config:

```nginx
location = /pubsub {
    nchan_pubsub websocket;
    nchan_channel_id foo;
    nchan_subscriber_first_message newest;
}
```

This works as expected, I get an echo message back. However when try to use redis with *nchan_redis_pass* or *nchan_use_redis* directive, the behavior is different:

1. When I send a message the client gets disconnected, however the next time I connect the message is delivered.
2. When I connect I receive all messages (regardless *nchan_subscriber_first_message newest*).

It appears with *nchan version <=1.1.4* the client is not disconnected whiling sending a message.

The behavior with `nchan_message_buffer_length 1;` is that on connect:

- after server restart: just one message received, setting to 0 delivers all messages.
- client disconnect/connect: no messages.

**Comment by slact** [comment]
Confirmed. I will have a fix within a few days.

**Comment by slact** [comment]
Fixed in 5744ea57. Also added missing tests for this issue in 05d6a762.

Thanks for the detailed bug report and your regression testing. You can test the fix in master.

**Comment by akornatskyy** [comment]
The directive `nchan_subscriber_first_message` is not respected after nginx restart.

1. `nchan_subscriber_first_message newest`:
1.1. disconnect / connect: as expected.
1.2. nginx restart: last 10 re-delivered; expected none.
2. `nchan_subscriber_first_message 0`: same behavior as with *newest*
3. `nchan_subscriber_first_message 1`:
3.1. disconnect / connect: last 10 re-delivered; expected none.
3.2. nginx restart: last 10 re-delivered; expected none.

**Comment by slact** [comment]
Looking into this.

**Comment by slact** [comment]
Tentative fix in master @ eda2667a -- it works, but still needs thorough testing.

**Comment by slact** [comment]
Fix passed my tests and was released in version 1.1.10. Closing.
